library/ziti_ctrl.c

```diff
@@ -0,0 +1,293 @@
+
+#define CODE_MAP(XX) \
+XX(NO_ROUTABLE_INGRESS_NODES, ZITI_GATEWAY_UNAVAILABLE) \
+XX(NO_EDGE_ROUTERS_AVAILABLE, ZITI_GATEWAY_UNAVAILABLE) \
+XX(INVALID_AUTHENTICATION, ZITI_NOT_AUTHORIZED) \
+XX(REQUIRES_CERT_AUTH, ZITI_NOT_AUTHORIZED)\
+XX(UNAUTHORIZED, ZITI_NOT_AUTHORIZED)\
+XX(INVALID_AUTH, ZITI_NOT_AUTHORIZED)
+
+#define CODE_MATCH(c, err) if (strcmp(code,#c) == 0) return err;
+
+    CODE_MAP(CODE_MATCH)
+
+    ZITI_LOG(WARN, "unmapped error code: %s", code);
+    return ZITI_WTF;
+}
+
+struct ctrl_resp {
+    int status;
+    char *body;
+    size_t received;
+    bool resp_chunked;
+
+    void* (*body_parse_func)(const char*, int);
+    void (*resp_cb)(void*, ziti_error*, void*);
+    void *ctx;
+
+    ziti_controller *ctrl;
+    void (*ctrl_cb)(void*, ziti_error*, struct ctrl_resp*);
+};
+
+static void ctrl_resp_cb(um_http_req_t *req, int code, um_header_list *headers) {
+    struct ctrl_resp *resp = req->data;
+    resp->status = code;
+    um_http_hdr *h;
+    LIST_FOREACH(h, headers, _next) {
+        if (strcasecmp(h->name, "Content-Length") == 0) {
+            resp->body = malloc(atol(h->value));
+            break;
+        }
+        if (strcasecmp(h->name, "Transfer-Encoding") == 0 && strcmp(h->value, "chunked") == 0) {
+            resp->resp_chunked = true;
+            resp->body = malloc(0);
+        }
+    }
+}
+
+static void ctrl_default_cb (void *s, ziti_error *e, struct ctrl_resp *resp) {
+    if (resp->resp_cb) {
+        resp->resp_cb(s, e, resp->ctx);
+    }
+
+    free(resp);
+}
+
+static void ctrl_login_cb(ziti_session *s, ziti_error *e, struct ctrl_resp *resp) {
+    resp->ctrl->session = strdup(s->token);
+    um_http_header(&resp->ctrl->client, "zt-session", s->token);
+    ctrl_default_cb(s, e, resp);
+}
+
+static void ctrl_logout_cb(void *s, ziti_error *e, struct ctrl_resp *resp) {
+    FREE(resp->ctrl->session);
+    um_http_header(&resp->ctrl->client, "zt-session", NULL);
+    ctrl_default_cb(s, e, resp);
+}
+
+static void ctrl_service_cb(ziti_service **services, ziti_error *e, struct ctrl_resp *resp) {
+    ziti_service *s = services != NULL ? services[0] : NULL;
+    ctrl_default_cb(s, e, resp);
+    free(services);
+}
+
+static void free_body_cb(um_http_req_t *req, const char *body, ssize_t len) {
+    free(body);
+}
+
+static void ctrl_body_cb(um_http_req_t *req, const char* b, ssize_t len) {
+    struct ctrl_resp *resp = req->data;
+
+    if (len > 0) {
+        if (resp->resp_chunked) {
+            resp->body = realloc(resp->body, resp->received + len);
+        }
+        memcpy(resp->body + resp->received, b, len);
+        resp->received += len;
+    } else if (len == UV_EOF) {
+        const char* data;
+        int data_len;
+        void *resp_obj = NULL;
+        ziti_error *err = NULL;
+
+        if (resp->status > 299) {
+            mjson_find(resp->body, resp->received, "$.error", (const char **) &data, &data_len);
+            err = parse_ziti_error(data, data_len);
```

Comment:
300's are 'errors'?

Reply:
we don't expect them since controller does not do redirects